presto-main/src/main/java/com/facebook/presto/execution/TaskStatus.java

```diff
@@ -0,0 +1,170 @@
+
+    @JsonCreator
+    public TaskStatus(@JsonProperty("taskId") TaskId taskId,
+            @JsonProperty("taskInstanceId") String taskInstanceId,
+            @JsonProperty("version") long version,
+            @JsonProperty("state") TaskState state,
+            @JsonProperty("self") URI self,
+            @JsonProperty("failures") List<ExecutionFailureInfo> failures,
+            @JsonProperty("queuedPartitionedDrivers") int queuedPartitionedDrivers,
+            @JsonProperty("runningPartitionedDrivers") int runningPartitionedDrivers,
+            @JsonProperty("memoryReservation") DataSize memoryReservation)
+    {
+        this.taskId = requireNonNull(taskId, "taskId is null");
+        this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceId is null");
+
+        this.version = version;
+        this.state = requireNonNull(state, "state is null");
+        this.self = requireNonNull(self, "self is null");
+
+        checkArgument(queuedPartitionedDrivers >= 0, "queuedPartitionedDrivers must be positive");
+        this.queuedPartitionedDrivers = queuedPartitionedDrivers;
+
+        checkArgument(runningPartitionedDrivers >= 0, "runningPartitionedDrivers must be positive");
+        this.runningPartitionedDrivers = runningPartitionedDrivers;
+
+        this.memoryReservation = requireNonNull(memoryReservation, "memoryReservation is null");
+        this.failures = (failures == null) ? ImmutableList.of() : ImmutableList.copyOf(failures);
+    }
+
+    @JsonProperty
+    public TaskId getTaskId()
+    {
+        return taskId;
+    }
+
+    @JsonProperty
+    public String getTaskInstanceId()
+    {
+        return taskInstanceId;
+    }
+
+    @JsonProperty
+    public long getVersion()
+    {
+        return version;
+    }
+
+    @JsonProperty
+    public TaskState getState()
+    {
+        return state;
+    }
+
+    @JsonProperty
+    public URI getSelf()
+    {
+        return self;
+    }
+
+    @JsonProperty
+    public List<ExecutionFailureInfo> getFailures()
+    {
+        return failures;
+    }
+
+    @JsonProperty
+    public int getQueuedPartitionedDrivers()
+    {
+        return queuedPartitionedDrivers;
+    }
+
+    @JsonProperty
+    public int getRunningPartitionedDrivers()
+    {
+        return runningPartitionedDrivers;
+    }
+
+    @JsonProperty
+    public DataSize getMemoryReservation()
+    {
+        return memoryReservation;
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("taskId", taskId)
+                .add("state", state)
+                .toString();
+    }
+
+    public static TaskStatus initialTaskStatus(TaskId taskId, URI location)
+    {
+        return new TaskStatus(taskId, "", MIN_VERSION, PLANNED, location, ImmutableList.of(), 0, 0, new DataSize(0, BYTE));
+    }
+
+    public static TaskStatus failWith(TaskStatus taskStatus, TaskState state, List<ExecutionFailureInfo> exceptions)
```

Comment:
Why is this a static method?  Also, is this ever called with a state other than `FAILED`?